lib/ecmascript.ts

```diff
@@ -917,87 +923,80 @@ export function ToSecondsStringPrecision(options: Temporal.ToStringPrecisionOpti
   }
 }
 
-type ToSingularUnit<T extends Temporal.DateTimeUnit | Temporal.PluralUnit<Temporal.DateTimeUnit> | 'auto'> = Exclude<
-  T,
-  Temporal.PluralUnit<Temporal.DateTimeUnit> | 'auto'
->;
+export const REQUIRED = Symbol('~required~');
 
-export function ToLargestTemporalUnit<Allowed extends Temporal.DateTimeUnit>(
-  options: { largestUnit?: Temporal.LargestUnit<Allowed> },
-  fallback: undefined
-): ToSingularUnit<Allowed> | 'auto' | undefined;
-export function ToLargestTemporalUnit<
-  Allowed extends Temporal.LargestUnit<Temporal.DateTimeUnit>,
-  Disallowed extends Temporal.DateTimeUnit
->(
-  options: { largestUnit?: Allowed | undefined },
-  fallback: 'auto',
-  disallowedStrings: ReadonlyArray<Disallowed>,
-  autoValue: ToSingularUnit<Allowed>
-): ToSingularUnit<Allowed>;
-export function ToLargestTemporalUnit<
-  Allowed extends Temporal.DateTimeUnit,
-  Disallowed extends ToSingularUnit<Exclude<Temporal.DateTimeUnit, Allowed>>,
-  Fallback extends ToSingularUnit<Allowed> | 'auto' | undefined
+interface TemporalUnitOptionsBag {
+  smallestUnit?: Temporal.PluralUnit<Temporal.DateTimeUnit> | Temporal.DateTimeUnit | 'auto';
+  largestUnit?: Temporal.PluralUnit<Temporal.DateTimeUnit> | Temporal.DateTimeUnit | 'auto';
+  unit?: Temporal.PluralUnit<Temporal.DateTimeUnit> | Temporal.DateTimeUnit | 'auto';
+}
+type UnitOptionsBagKeys = keyof TemporalUnitOptionsBag;
+type UnitTypeMapping = {
+  date: Temporal.DateUnit;
+  time: Temporal.TimeUnit;
+  datetime: Temporal.DateTimeUnit;
+};
+
+export function GetTemporalUnit<
+  T extends keyof UnitTypeMapping,
+  D extends typeof REQUIRED | UnitTypeMapping[T] | 'auto' | undefined,
+  R extends Exclude<D, typeof REQUIRED> | UnitTypeMapping[T]
```

Comment:
If I'm reading this correctly, `Exclude<D, typeof REQUIRED>` already is unioned with `UnitTypeMapping[T]`, so that might be redundant?

Reply:
I don't believe it's redundant because of the way that the binding D happens.

If I write

`GetTemporalUnit(o, 'smallestUnit', 'time', 'hours')`

then `D` is bound to `'hours'`, not to the whole union type space.

Therefore, we need to re-add that space back into the return type (as the allowed values also include all `TimeUnit` values).